File: apps/desktop/src/shared/utils/schedule.ts
// Scheduling helpers used by both calendar blocks and the editor's
// date-picker chip. Pure date-string transforms — no calendar layout deps —
// so they belong in shared/ rather than features/calendar/utils.

import { formatLocalISO, isAllDayIso, isTimedIso, parseLocalISO } from "@pikos/core";

/**
 * Decides what to do with an event's end when the user picks a new start ISO
 * in the date picker. Handles four transitions:
 *   - all-day → timed: collapse to single day (end undefined)
 *   - timed → all-day: preserve date extent (drop time from end)
 *   - timed → timed: preserve duration
 *   - all-day → all-day: preserve end, drop if it's now before the new start
 * Returns start = iso and either an end ISO or undefined (single occurrence).
 * `iso` must be non-null — callers should treat null as "clear schedule" first.
 */
export function computeScheduleTransition(
  current: { start: string | null | undefined; end: string | null | undefined },
  iso: string
): { start: string; end: string | undefined } {
  const currStart = current.start;
  const currEnd = current.end;
  const wasAllDay = currStart != null && isAllDayIso(currStart);
  const nowAllDay = isAllDayIso(iso);

  // all-day → timed: collapse to single day.
  if (wasAllDay && !nowAllDay) return { end: undefined, start: iso };

  // timed → all-day: strip time from end; preserve date extent.
  if (currStart != null && !wasAllDay && nowAllDay) {
    if (currEnd && isTimedIso(currEnd)) {
      const endDateOnly = currEnd.slice(0, 10);
      return { end: endDateOnly > iso ? endDateOnly : undefined, start: iso };
    }
    return { end: currEnd && currEnd > iso ? currEnd : undefined, start: iso };
  }

  // timed → timed: preserve duration.
  if (
    !nowAllDay &&
    currStart != null &&
    isTimedIso(currStart) &&
    currEnd != null &&
    isTimedIso(currEnd)
  ) {
    const durationMs = parseLocalISO(currEnd).getTime() - parseLocalISO(currStart).getTime();
    if (durationMs > 0) {
      const endIso = formatLocalISO(new Date(parseLocalISO(iso).getTime() + durationMs));
      return { end: endIso, start: iso };
    }
    return { end: currEnd, start: iso };
  }

  // all-day → all-day: preserve end, drop if now before new start.
  return { end: currEnd && currEnd >= iso ? currEnd : undefined, start: iso };
}

/**
 * Normalises the value returned from the end-date picker into the `end` arg
 * for `scheduleOnce`. Returns `undefined` when the picker cleared the end or
 * when the resulting end would be <= start (single-day semantics). For all-day
 * starts, any datetime end is stripped back to date-only.
 */
export function normalizeEndInput(currentStart: string, endIso: string | null): string | undefined {
  if (endIso === null) return undefined;
  const startIsAllDay = isAllDayIso(currentStart);
  let next = endIso;
  if (startIsAllDay && isTimedIso(next)) next = next.slice(0, 10);
  if (next <= currentStart) return undefined;
  return next;
}

File: crates/pikos-core/src/schedule.rs
//! All-day ↔ timed schedule transitions.
//!
//! Port of `apps/desktop/src/shared/utils/schedule.ts`. These decide what
//! happens to an event's *end* when the user picks a new start in a date
//! picker. A mobile date picker faces exactly the same four transitions, and
//! disagreeing about them would mean the same edit produces different events
//! depending on which device made it.

use chrono::Duration;

use crate::dates::{format_local_iso, is_all_day_iso, is_timed_iso, parse_local_iso};

/// The resulting schedule. `end` of `None` means a single occurrence.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ScheduleTransition {
    pub start: String,
    pub end: Option<String>,
}

/// Decide the new end when the start moves to `iso`.
///
/// Four transitions, in the order the original tests them:
///
/// - all-day → timed: collapse to a single occurrence
/// - timed → all-day: keep the date extent, drop the time
/// - timed → timed: preserve duration
/// - all-day → all-day: keep the end unless it now precedes the start
pub fn compute_schedule_transition(
    current_start: Option<&str>,
    current_end: Option<&str>,
    iso: &str,
) -> ScheduleTransition {
    let was_all_day = current_start.is_some_and(is_all_day_iso);
    let now_all_day = is_all_day_iso(iso);

    // all-day → timed.
    if was_all_day && !now_all_day {
        return ScheduleTransition {
            start: iso.to_string(),
            end: None,
        };
    }

    // timed → all-day.
    if current_start.is_some() && !was_all_day && now_all_day {
        if let Some(end) = current_end.filter(|e| is_timed_iso(e)) {
            let date_only = &end[..10.min(end.len())];
            return ScheduleTransition {
                start: iso.to_string(),
                end: (date_only > iso).then(|| date_only.to_string()),
            };
        }
        return ScheduleTransition {
            start: iso.to_string(),
            end: current_end
                .filter(|e| !e.is_empty() && *e > iso)
                .map(str::to_string),
        };
    }

    // timed → timed.
    if !now_all_day {
        if let (Some(start), Some(end)) = (current_start, current_end) {
            if is_timed_iso(start) && is_timed_iso(end) {
                if let (Some(s), Some(e), Some(new_start)) = (
                    parse_local_iso(start),
                    parse_local_iso(end),
                    parse_local_iso(iso),
                ) {
                    let duration = e - s;
                    if duration > Duration::zero() {
                        return ScheduleTransition {
                            start: iso.to_string(),
                            end: Some(format_local_iso(&(new_start + duration))),
                        };
                    }
                }
                // Zero or negative duration keeps the existing end verbatim,
                // matching the original's fall-through.
                return ScheduleTransition {
                    start: iso.to_string(),
                    end: Some(end.to_string()),
                };
            }
        }
    }

    // all-day → all-day.
    ScheduleTransition {
        start: iso.to_string(),
        end: current_end
            .filter(|e| !e.is_empty() && *e >= iso)
            .map(str::to_string),
    }
}

/// Normalise an end-date picker result.
///
/// `None` when the picker cleared the end, or when the end would not be after
/// the start — both mean a single-day event. An all-day start strips a timed
/// end back to its date.
pub fn normalize_end_input(current_start: &str, end_iso: Option<&str>) -> Option<String> {
    let end = end_iso?;
    let mut next = end.to_string();
    if is_all_day_iso(current_start) && is_timed_iso(&next) {
        next.truncate(10);
    }
    if next.as_str() <= current_start {
        return None;
    }
    Some(next)
}
